ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -98,6 +104,36 @@ data ServerState txid tx = ServerState {
        -- transactions out of order but must use the original order when adding
        -- to the mempool or acknowledging transactions.
        --
+       -- However, it's worth noting that, in a few situations, some of the
+       -- transaction IDs in this 'Map' may be mapped to 'Nothing':
+       --
+       -- * Transaction IDs mapped to 'Nothing' can represent transaction IDs
+       --   that were requested, but not received. This can occur because the
+       --   client will not necessarily send all of the transactions that we
+       --   asked for, but we still need to acknowledge those transactions.
+       --
+       --   For example, if we request a transaction that no longer exists in
+       --   the client's mempool, the client will just exclude it from the
+       --   response. However, we still need to acknowledge it (i.e. remove it
+       --   from the 'unacknowledgedTxIds') in order to note that we're no
+       --   longer awaiting receipt of that transaction.
+       --
+       -- * Transaction IDs mapped to 'Nothing' can represent transactions that
+       --   were not requested from the client because they already exist
+       --   within the mempool.
+       --
+       --   TODO @intricate: The paragraph below is slightly outdated as I've
+       --                    made some changes since. Don't forget to update
+       --                    it.
+       --
+       --   For example, if we request some transaction IDs and otice that
+       --   some subset of them already exist within our mempool, we wouldn't
+       --   want to bother asking for those specific transactions since we
+       --   already have them. Therefore, we would just add those transaction
+       --   IDs mapped to 'Nothing' to the 'bufferedTxs' such that those
+       --   transactions are eventually acknowledged, but never actually
+       --   requested.
+       --
        bufferedTxs         :: Map txid (Maybe tx),
```

Comment:
This record is the state of this server, so we should check whether this contains thunks to avoid space leaks. This record is not stored in a `StrictTVar` (or variant), but passed around. See https://github.com/input-output-hk/ouroboros-network/blob/5c90d6e6e57f46f7fe22aa5fd657554493a71948/ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs#L786 for an example on how to check for thunks in such a state. Slightly annoying: this code lives in `ouroboros-network`, while the thunks checking code lives in consensus. So something should be moved at some point.

This is orthogonal to this PR, so let's create a separate issue for this.

Reply:
Yup: https://github.com/input-output-hk/ouroboros-network/issues/1645